engine/runner/src/main/java/org/enso/runner/Main.java

```diff
@@ -0,0 +1,1413 @@
+            .build();
+
+    var warningsLimitOption =
+        cliOptionBuilder()
+            .longOpt(WARNINGS_LIMIT)
+            .hasArg(true)
+            .numberOfArgs(1)
+            .argName("limit")
+            .desc("Specifies a maximal number of reported warnings. Defaults to `100`.")
+            .build();
+
+    var disablePrivateCheckOption =
+        cliOptionBuilder()
+            .longOpt(DISABLE_PRIVATE_CHECK_OPTION)
+            .desc("Disables private module checking at runtime. Useful for tests.")
+            .build();
+
+    var options = new Options();
+    options
+        .addOption(help)
+        .addOption(repl)
+        .addOption(run)
+        .addOption(inspect)
+        .addOption(dumpGraphs)
+        .addOption(docs)
+        .addOption(preinstall)
+        .addOption(newOpt)
+        .addOption(newProjectNameOpt)
+        .addOption(newProjectModuleNameOpt)
+        .addOption(newProjectTemplateOpt)
+        .addOption(newProjectAuthorNameOpt)
+        .addOption(newProjectAuthorEmailOpt)
+        .addOption(lsOption)
+        .addOption(lsProfilingPathOption)
+        .addOption(lsProfilingTimeOption)
+        .addOption(deamonizeOption)
+        .addOption(interfaceOption)
+        .addOption(rpcPortOption)
+        .addOption(dataPortOption)
+        .addOption(secureRpcPortOption)
+        .addOption(secureDataPortOption)
+        .addOption(uuidOption)
+        .addOption(pathOption)
+        .addOption(inProjectOption)
+        .addOption(version)
+        .addOption(json)
+        .addOption(logLevelOption)
+        .addOption(loggerConnectOption)
+        .addOption(noLogMaskingOption)
+        .addOption(uploadOption)
+        .addOption(updateManifestOption)
+        .addOption(hideProgressOption)
+        .addOption(authTokenOption)
+        .addOption(noReadIrCachesOption)
+        .addOption(compileOption)
+        .addOption(noCompileDependenciesOption)
+        .addOption(noGlobalCacheOption)
+        .addOptionGroup(cacheOptionsGroup)
+        .addOption(autoParallelism)
+        .addOption(skipGraalVMUpdater)
+        .addOption(executionEnvironmentOption)
+        .addOption(warningsLimitOption)
+        .addOption(disablePrivateCheckOption);
+
+    return options;
+  }
+
+  /**
+   * Prints the help message to the standard output.
+   *
+   * @param options object representing the CLI syntax
+   */
+  private static void printHelp(Options options) {
+    new HelpFormatter().printHelp(LanguageInfo.ID, options);
+  }
+
+  /** Terminates the process with a failure exit code. */
+  private RuntimeException exitFail() {
+    return doExit(1);
+  }
+
+  /** Terminates the process with a success exit code. */
+  private RuntimeException exitSuccess() {
+    return doExit(0);
+  }
+
+  /** Shuts down the logging service and terminates the process. */
+  private RuntimeException doExit(int exitCode) {
+    RunnerLogging.tearDown();
+    System.exit(exitCode);
+    return null;
```

Comment:
That and phantom "throwing" feels like a really weird pattern overall.

Reply:
It is necessary to write `throw exitFailure()` otherwise Java requires a `return` value or initialization of a local variable defined sooner, but used after the _exit_ statement.